### ISSUE TITLE
Native MultiMC pack format

### ISSUE THREAD

**Issue by peterix** [MultiMC/MultiMC5#916]
We need one.

The following transcript is full of helpful hints:

**Peterix:**
I'll be expanding upon the pack support in the semi-near future. Hopefully with something that can grab updates too.
ftb/curse/technic/whatever... but it will take time

**Heph:**
i just took about 30 seconds and already have a scheme for easy updating using xml
for packs that arent hosted by an external launcher

**Peterix:**
there really isn't that much to it, yeah

**Heph:**
not sure how to handle forge version changes since its hosted outside the minecraft folder in a json and loaded dynamically
the only thing i could thikn of would be that the xml hosted remotely would have to specify the forge version number
in a tag:

```
<forge>
<version>x.y.z.xxx</version>
</forge>
```

**Peterix:**
if I keep just the dumb big zip files, it's a list of links to dumb big zip files, with 'latest' and 'recommended' somehow specified

**Heph:**
instance.cfg would have to have two additional keys added to it

**Peterix:**
maybe a link to some image and a description
I'd also like to add icons embedded in instances
and some more memory settings options, friendlier to distribution

**Heph:**
pack version, and remote xml
in the remote xml, forge, pack version, and the path to the update file to be downloaded/applied
the format of the update zip would be rather straightforward
would need to also address deletions or changed filenames or what have you
like if an update removed a mod, you want it deleted locally on update

**Peterix:**
yeah. I do something similar with mmc updates already
it compares files in current and previous versions and determines what to do

**Heph:**
could just have a <removed> <file></file></removed> section
any filename in a file tag is searched for in /mods and removed
easy peasy

**Peterix:**
I'd just list the files and let the thing figure out what to do... just needs to store what was actually installed, with checksums. so you can update
I mean, this is actually in mmc code already. just only used for its own updates

**Heph:**
make the pack maker do all the work
keep it simple on MultiMC's side :P
so much extra complexity and programming, when you can just specify an XML format schema and have it be strict
and blame it on bad XML :P

**Peterix:**
lol. I guess :)

**Heph:**
directory searches arent compute heavy
takes miliseconds to parse the entire mods list

**Peterix:**
well, if you go into every single thing in there and hash it... then it can take a bit more
but yeah
the downloads will take longer anyway

**Heph:**
i dont think you need to hash things :P verify the integrity of the update against an md5 stored in the XML, and you can guarantee everything is kosher

**Peterix:**
well, if there's anything I learned from this so far, people are really good at breaking things

**Heph:**
make it really hard to break things.
if you want ill draft up a sample XML file for you to fiddle with :P

**Peterix:**
sure. I might end up with something simpler/more stupid though. or more json-like, because that's what the mod metadata repo thingy will probably use

**Heph:**
i know someone with a massive collection of mods, and who already has the entire archive with md5+sha1 checksums :P
im working on adding metadata about every file (mc version, forge version) but it takes time

**Peterix:**
yeah. even if I can't distribute the mod itself, knowing what it is and having the checksums is really good
means the launcher can identify it and tell you where to grab updates

**Heph:**
the database holding all the checksums is 1/5 the size of teh archive :P

**Peterix:**
http://wonko.multimc.org/org.lwjgl.json
http://wonko.multimc.org/org.lwjgl/2.9.4-nightly-20150209.json

**Heph:**
JSON is just about the worst when it comes to data efficiency :P
such a waste of all that bytes

**Peterix:**
yeah. but it's good enough for a proof of concept :P

**Heph:**
but JSON is extremely easy to generate automatically
and thats where its strength is

**Peterix:**
and then you write down a schema, generate parsers and serializers with protobuf (or similar) and compress the hell out of it
grand plans, but too big of a leap

**Heph:**
take small steps :P

**Peterix:**
I do. :D
I'll put this in a github issue if you don't mind :)

**Heph:**
sure :)

**Comment by 02JanDal** [comment]
Well, I'd like packs to just be another wonko version. So a pack could look like this:

```json
{
  "formatVersion": 0,
  "uid": "ftb.infinity",
  "version": "1.2.3.4",
  "time": 1417952120,
  "type": "release",
  "data": [
    {
      "rules": [
        { "action": "disallow" },
        { "action": "allow", "side": "client" }
      ],
      "requires": [
        { "uid": "net.minecraftforge", "version": "11.14.0.1257" },
        { "uid": "buildcraft", "version": "7.42" }
      ],
      "memory": {
        "min": 1024,
        "max": 9999
      }
    }
  ]
}
```

Some things to note: I made `requires` a resource and added `memory`. Doing it like this means we can reuse parsing and other logic, like sidedness.

The question here is what happens if the user edits (adds/removes something to/from) the pack. When that happens (or already when first creating the instance?) a file like this is the "actual" instance pack:

```json
{
  "formatVersion": 0,
  "uid": "ftb.infinity",
  "version": "1.2.3.4-custom",
  "time": 1417952120,
  "type": "release",
  "data": [
    {
      "rules": [
        { "action": "disallow" },
        { "action": "allow", "side": "client" }
      ],
      "requires": [
        { "uid": "ftb.infinity", "version": "1.2.3.4" },
        { "uid": "my.custom.added.mod", "version": "x.y.z" }
      ]
    }
  ]
}
```

The user could then simply release his modified pack by distributing that JSON, which itself also is a pack. Collaboration :D

Basically what I'm saying is: don't invent another format, use the existing one.

An open question (both for this issue and wonko generally) is what we want to use instead of JSON (that we want to use something different already seems to be pretty much decided).

**Comment by peterix** [comment]
I see modpacks as something orthogonal to the metadata repo - they are their own thing. And in MultiMC, they should be their own thing, completely separate. The differences between pack formats are too big to shoehorn them into one unified format without repackaging and redistribution.

**Comment by 02JanDal** [comment]
What differences?

**Comment by peterix** [comment]
A curse pack is a zip with part of the data and metadata inside - missing mods, but including 'overrides' that can be *anything*. Patches are completely undefined and they use installers.
A technic pack is a bunch of metadata on a server and zips within zips, with weird rules about how to extract them right. Also has its own patch files.
FTB? More weird stuff, and their own patch file format.
AT launcher, I don't even know... it's a bunch of weird stuff.

**Comment by peterix** [comment]
Our own? An instance dump.
I would be for the curse model, because it's the least complicated, but using wonko for packages.

**Comment by 02JanDal** [comment]
Well sure, that would allow easily adding custom mods without having them in wonko. But then the metadata file (manifest.json in the case of curse, possibly something else in our case) could still have the above format.

Pack format == instance format == wonko format

**Comment by peterix** [comment]
Sure. that sounds good.

**Comment by 02JanDal** [comment]
I'm kinda against allowing older instance formats to be made into packs. It would increase their time of life and would be harder for us to support. Also means it would be harder for others to use.

**Comment by peterix** [comment]
On the other hand, that's what we have now, and people already use it.

**Comment by 02JanDal** [comment]
Kind of sadly. The point about harder for others still stands though.

**Comment by peterix** [comment]
In the name of small steps, I moved this to [workflowy (Ideas -> Packs)](https://workflowy.com/s/2EyDMcp7CU), so I can make more small steps in a smaller yard, while not having to scale any walls.
